query/stdlib/influxdata/influxdb/provider.go

```diff
@@ -0,0 +1,166 @@
+	"github.com/influxdata/flux/codes"
+	"github.com/influxdata/flux/dependencies/influxdb"
+	"github.com/influxdata/flux/execute"
+	"github.com/influxdata/flux/memory"
+	"github.com/influxdata/flux/values"
+	"github.com/influxdata/influxdb/v2/kit/platform"
+	"github.com/influxdata/influxdb/v2/kit/platform/errors"
+	"github.com/influxdata/influxdb/v2/query"
+	"github.com/influxdata/influxdb/v2/storage/reads/datatypes"
+)
+
+type (
+	Config       = influxdb.Config
+	Predicate    = influxdb.Predicate
+	PredicateSet = influxdb.PredicateSet
+)
+
+// Provider is an implementation of influxdb.Provider that exposes the
+// query.StorageReader to flux and, if a host or org were specified, it
+// delegates to the influxdb.HttpProvider.
+type Provider struct {
+	influxdb.HttpProvider
+	Reader       query.StorageReader
+	BucketLookup BucketLookup
+}
+
+func (p Provider) SeriesCardinalityReaderFor(ctx context.Context, conf influxdb.Config, bounds flux.Bounds, predicateSet influxdb.PredicateSet) (influxdb.Reader, error) {
+	// If an organization is specified, it must be retrieved through the http
+	// provider.
+	if conf.Org.IsValid() || conf.Host != "" {
+		return p.HttpProvider.SeriesCardinalityReaderFor(ctx, conf, bounds, predicateSet)
+	}
+
+	if !p.Reader.SupportReadSeriesCardinality(ctx) {
+		return nil, &errors.Error{
+			Code: errors.EInvalid,
+			Msg:  "series cardinality option is not supported",
+		}
+	}
+
+	spec, err := p.readFilterSpec(ctx, conf, bounds, predicateSet)
+	if err != nil {
+		return nil, err
+	}
+	return seriesCardinalityReader{
+		reader: p.Reader,
+		spec: query.ReadSeriesCardinalitySpec{
+			ReadFilterSpec: spec,
+		},
+	}, nil
+}
+
+// readFilterSpec will construct a query.ReadFilterSpec from the context and the
+// configuration parameters.
+func (p Provider) readFilterSpec(ctx context.Context, conf influxdb.Config, bounds flux.Bounds, predicateSet influxdb.PredicateSet) (query.ReadFilterSpec, error) {
+	// Retrieve the organization id from the request context. Do not use the
+	// configuration.
+	req := query.RequestFromContext(ctx)
+	if req == nil {
+		return query.ReadFilterSpec{}, &errors.Error{
+			Code: errors.EInvalid,
+			Msg:  "missing request on context",
+		}
+	}
+
+	orgID := req.OrganizationID
+	bucketID, err := p.lookupBucketID(ctx, orgID, conf.Bucket)
+	if err != nil {
+		return query.ReadFilterSpec{}, err
+	}
+
+	spec := query.ReadFilterSpec{
+		OrganizationID: orgID,
+		BucketID:       bucketID,
+		Bounds: execute.Bounds{
+			Start: values.ConvertTime(bounds.Start.Time(bounds.Now)),
+			Stop:  values.ConvertTime(bounds.Stop.Time(bounds.Now)),
+		},
+	}
+
+	if len(predicateSet) > 0 {
+		predicates := make([]*datatypes.Predicate, 0, len(predicateSet))
+		for _, predicate := range predicateSet {
+			fn, ok := predicate.Fn.GetFunctionBodyExpression()
+			if !ok {
+				return query.ReadFilterSpec{}, &flux.Error{
+					Code: codes.Invalid,
+					Msg:  "predicate body cannot be pushed down",
```

Comment:
Do we need to error in this case because it's infeasible to compute the cardinality in the flux engine?

Reply:
This was some "leveraged" code, but from the way it looks to me: Yes, cardinality is determined entirely outside of the flux engine, and this is kind of an extra check to make sure that the predicate is valid for using in storage. I don't honestly know how one would create a predicate that would trigger this error, but that's probably just a lack of imagination on my part.

That said, the actual storage code to come will certainly also return errors if the predicate can't be interpreted for any reason.